File: backend/app/modules/inventory/_qty_ledger.py
"""Shared batch quantity helpers — deduct/restore + event writing."""
import datetime
from decimal import Decimal
from typing import Optional

from fastapi import HTTPException
from sqlalchemy.orm import Session

_BLOCKED_STATUSES = {"CONSUMED", "EXPIRED", "QUARANTINE"}


def _deduct_from_packs(db: Session, batch_id: int, qty: Decimal) -> None:
    """Apply a whole-batch deduction across a batch's packs (FIFO by seq_no)
    so pack-level qty_available stays reconciled with the batch total."""
    from app.models.inventory import InvBatchPack

    remaining = qty
    packs = (
        db.query(InvBatchPack)
        .filter(InvBatchPack.batch_id == batch_id, InvBatchPack.qty_available > 0)
        .order_by(InvBatchPack.seq_no)
        .all()
    )
    for pack in packs:
        if remaining <= 0:
            break
        take = min(pack.qty_available, remaining)
        pack.qty_available = pack.qty_available - take
        remaining -= take


def _restore_to_packs(db: Session, batch_id: int, qty: Decimal) -> None:
    """Return a whole-batch restore across a batch's packs (fill least-full first,
    capped at each pack's original qty_per_pack)."""
    from app.models.inventory import InvBatchPack

    remaining = qty
    packs = (
        db.query(InvBatchPack)
        .filter(InvBatchPack.batch_id == batch_id)
        .order_by(InvBatchPack.seq_no)
        .all()
    )
    for pack in packs:
        if remaining <= 0:
            break
        room = (pack.qty_per_pack or Decimal("0")) - (pack.qty_available or Decimal("0"))
        if room <= 0:
            continue
        give = min(room, remaining)
        pack.qty_available = (pack.qty_available or Decimal("0")) + give
        remaining -= give


def deduct_qty(
    db: Session,
    batch_id: int,
    qty: Decimal,
    *,
    event_type: str,
    performed_by: str,
    ref_no: Optional[str] = None,
    module: Optional[str] = None,
    issued_to: Optional[str] = None,
    purpose: Optional[str] = None,
    project_code: Optional[str] = None,
    remarks: Optional[str] = None,
) -> None:
    """Deduct qty from batch, update status, write a batch event row."""
    from app.models.inventory import InvBatch, InvBatchEvent

    batch = db.get(InvBatch, batch_id)
    if not batch:
        raise HTTPException(404, f"Batch {batch_id} not found.")
    if batch.status in _BLOCKED_STATUSES or not batch:
        raise HTTPException(400, f"Batch {batch_id} has status '{batch.status}' and cannot be issued.")

    available = batch.qty_available or Decimal("0")
    if qty <= 0:
        raise HTTPException(400, "Quantity must be greater than zero.")
    if qty > available:
        raise HTTPException(
            400,
            f"Insufficient stock. Available: {available}, requested: {qty}.",
        )

    batch.qty_available = available - qty
    if batch.qty_available == 0:
        batch.status = "CONSUMED"
        batch.category = "historic"
    else:
        batch.status = "PARTIALLY_CONSUMED"

    if batch.include_pack:
        _deduct_from_packs(db, batch_id, qty)

    db.add(InvBatchEvent(
        batch_id=batch_id,
        event_type=event_type,
        qty=qty,
        ref_no=ref_no,
        module=module,
        issued_to=issued_to,
        purpose=purpose,
        project_code=project_code,
        performed_by=performed_by,
        performed_at=datetime.datetime.utcnow(),
        remarks=remarks,
    ))


def deduct_pack_qty(
    db: Session,
    pack_id: int,
    qty: Decimal,
    *,
    event_type: str,
    performed_by: str,
    ref_no: Optional[str] = None,
    module: Optional[str] = None,
    issued_to: Optional[str] = None,
    purpose: Optional[str] = None,
    project_code: Optional[str] = None,
    remarks: Optional[str] = None,
) -> None:
    """Deduct qty from a specific pack's own quantity (not the whole batch),
    keeping the parent batch's aggregate qty_available in sync, and write a
    batch event row. Use this whenever the consumer picked one specific
    SKU/Pack ID rather than the batch as a whole.
    """
    from app.models.inventory import InvBatch, InvBatchPack, InvBatchEvent

    pack = db.get(InvBatchPack, pack_id)
    if not pack:
        raise HTTPException(404, f"Pack {pack_id} not found.")
    batch = db.get(InvBatch, pack.batch_id)
    if not batch:
        raise HTTPException(404, f"Parent batch for pack {pack_id} not found.")
    if batch.status in _BLOCKED_STATUSES:
        raise HTTPException(400, f"Batch {batch.id} has status '{batch.status}' and cannot be issued.")

    pack_available = pack.qty_available if pack.qty_available is not None else Decimal("0")
    if qty <= 0:
        raise HTTPException(400, "Quantity must be greater than zero.")
    if qty > pack_available:
        raise HTTPException(
            400,
            f"Insufficient stock in pack '{pack.inhouse_batch_no}'. Available: {pack_available}, requested: {qty}.",
        )

    pack.qty_available = pack_available - qty

    batch_available = batch.qty_available or Decimal("0")
    batch.qty_available = max(batch_available - qty, Decimal("0"))
    if batch.qty_available == 0:
        batch.status = "CONSUMED"
        batch.category = "historic"
    else:
        batch.status = "PARTIALLY_CONSUMED"

    db.add(InvBatchEvent(
        batch_id=batch.id,
        event_type=event_type,
        qty=qty,
        ref_no=ref_no,
        module=module,
        issued_to=issued_to,
        purpose=purpose,
        project_code=project_code,
        performed_by=performed_by,
        performed_at=datetime.datetime.utcnow(),
        remarks=remarks,
    ))


def restore_qty(
    db: Session,
    batch_id: int,
    qty: Decimal,
    *,
    performed_by: str,
    remarks: Optional[str] = None,
) -> None:
    """Return qty to batch (e.g. on allocation cancel) and write ADJUSTMENT event."""
    from app.models.inventory import InvBatch, InvBatchEvent

    batch = db.get(InvBatch, batch_id)
    if not batch:
        raise HTTPException(404, f"Batch {batch_id} not found.")

    batch.qty_available = (batch.qty_available or Decimal("0")) + qty
    if batch.status in ("CONSUMED", "PARTIALLY_CONSUMED"):
        batch.status = "PARTIALLY_CONSUMED" if batch.qty_available < batch.qty_received else "AVAILABLE"
        batch.category = "available"

    if batch.include_pack:
        _restore_to_packs(db, batch_id, qty)

    db.add(InvBatchEvent(
        batch_id=batch_id,
        event_type="ADJUSTMENT",
        qty=qty,
        performed_by=performed_by,
        performed_at=datetime.datetime.utcnow(),
        remarks=remarks,
    ))

File: backend-node/src/utils/qtyLedger.ts
/**
 * Batch quantity ledger — port of backend/app/modules/inventory/_qty_ledger.py.
 * FIFO deduct by pack seq_no; restore fills least-full packs first.
 */
import { Transaction } from 'sequelize'
import { Op } from 'sequelize'
import { InvBatch, InvBatchEvent, InvBatchPack } from '../models/InventoryModels.model'
import { BadRequestError, NotFoundError } from './errors'

export const BLOCKED_STATUSES = new Set(['CONSUMED', 'EXPIRED', 'QUARANTINE'])

export async function deductFromPacks(batchId: number, qty: number, transaction?: Transaction): Promise<void> {
  let remaining = qty
  const packs = await InvBatchPack.findAll({
    where: { batchId, qtyAvailable: { [Op.gt]: 0 } },
    order: [['seqNo', 'ASC']],
    transaction,
    lock: transaction ? true : undefined,
  })
  for (const pack of packs) {
    if (remaining <= 0) break
    const take = Math.min(Number(pack.qtyAvailable), remaining)
    await pack.update({ qtyAvailable: Number(pack.qtyAvailable) - take }, { transaction })
    remaining -= take
  }
}

export async function restoreToPacks(batchId: number, qty: number, transaction?: Transaction): Promise<void> {
  let remaining = qty
  const packs = await InvBatchPack.findAll({
    where: { batchId },
    order: [['seqNo', 'ASC']],
    transaction,
    lock: transaction ? true : undefined,
  })
  for (const pack of packs) {
    if (remaining <= 0) break
    const room = Number(pack.qtyPerPack || 0) - Number(pack.qtyAvailable || 0)
    if (room <= 0) continue
    const give = Math.min(room, remaining)
    await pack.update({ qtyAvailable: Number(pack.qtyAvailable || 0) + give }, { transaction })
    remaining -= give
  }
}

function applyConsumedStatus(available: number): { qtyAvailable: number; status: string; category?: string } {
  // Matches Python deduct_qty: remaining 0 → CONSUMED/historic, else PARTIALLY_CONSUMED.
  if (available <= 0) {
    return { qtyAvailable: 0, status: 'CONSUMED', category: 'historic' }
  }
  return { qtyAvailable: available, status: 'PARTIALLY_CONSUMED' }
}

export async function deductQty(opts: {
  batchId: number
  qty: number
  eventType: string
  performedBy: string
  refNo?: string | null
  module?: string | null
  issuedTo?: string | null
  purpose?: string | null
  projectCode?: string | null
  remarks?: string | null
  transaction?: Transaction
}): Promise<{ qtyAvailable: number; status: string }> {
  const { batchId, qty, transaction } = opts
  const batch = await InvBatch.findByPk(batchId, { transaction, lock: transaction ? true : undefined })
  if (!batch) throw new NotFoundError('Batch not found')
  if (BLOCKED_STATUSES.has(batch.status ?? '')) {
    throw new BadRequestError(`Batch ${batchId} has status '${batch.status}' and cannot be issued.`)
  }
  if (qty <= 0) throw new BadRequestError('Quantity must be greater than zero.')
  const available = Number(batch.qtyAvailable || 0)
  if (qty > available) {
    throw new BadRequestError(`Insufficient stock. Available: ${available}, requested: ${qty}.`)
  }

  const next = applyConsumedStatus(available - qty)
  await batch.update({ qtyAvailable: next.qtyAvailable, status: next.status, ...(next.category ? { category: next.category } : {}), updatedAt: new Date() }, { transaction })
  if (batch.includePack) await deductFromPacks(batchId, qty, transaction)

  await InvBatchEvent.create({
    batchId,
    eventType: opts.eventType,
    qty,
    refNo: opts.refNo ?? null,
    module: opts.module ?? null,
    issuedTo: opts.issuedTo ?? null,
    purpose: opts.purpose ?? null,
    projectCode: opts.projectCode ?? null,
    performedBy: opts.performedBy,
    performedAt: new Date(),
    remarks: opts.remarks ?? null,
  }, { transaction })

  return { qtyAvailable: next.qtyAvailable, status: next.status }
}

/** Deduct from one pack and keep the parent batch aggregate in sync (Python deduct_pack_qty). */
export async function deductPackQty(opts: {
  packId: number
  qty: number
  eventType: string
  performedBy: string
  refNo?: string | null
  module?: string | null
  issuedTo?: string | null
  purpose?: string | null
  projectCode?: string | null
  remarks?: string | null
  transaction?: Transaction
}): Promise<{ qtyAvailable: number; status: string }> {
  const { packId, qty, transaction } = opts
  const pack = await InvBatchPack.findByPk(packId, { transaction, lock: transaction ? true : undefined })
  if (!pack) throw new NotFoundError('Pack not found')
  const batch = await InvBatch.findByPk(pack.batchId, { transaction, lock: transaction ? true : undefined })
  if (!batch) throw new NotFoundError('Parent batch for pack not found')
  if (BLOCKED_STATUSES.has(batch.status ?? '')) {
    throw new BadRequestError(`Batch ${batch.id} has status '${batch.status}' and cannot be issued.`)
  }
  if (qty <= 0) throw new BadRequestError('Quantity must be greater than zero.')
  const packAvailable = Number(pack.qtyAvailable ?? 0)
  if (qty > packAvailable) {
    throw new BadRequestError(
      `Insufficient stock in pack '${pack.inhouseBatchNo}'. Available: ${packAvailable}, requested: ${qty}.`,
    )
  }

  await pack.update({ qtyAvailable: packAvailable - qty }, { transaction })
  const next = applyConsumedStatus(Math.max(Number(batch.qtyAvailable || 0) - qty, 0))
  await batch.update({
    qtyAvailable: next.qtyAvailable,
    status: next.status,
    ...(next.category ? { category: next.category } : {}),
    updatedAt: new Date(),
  }, { transaction })

  await InvBatchEvent.create({
    batchId: batch.id,
    eventType: opts.eventType,
    qty,
    refNo: opts.refNo ?? null,
    module: opts.module ?? null,
    issuedTo: opts.issuedTo ?? null,
    purpose: opts.purpose ?? null,
    projectCode: opts.projectCode ?? null,
    performedBy: opts.performedBy,
    performedAt: new Date(),
    remarks: opts.remarks ?? null,
  }, { transaction })

  return { qtyAvailable: next.qtyAvailable, status: next.status }
}

export async function restoreQty(opts: {
  batchId: number
  qty: number
  performedBy: string
  remarks?: string | null
  transaction?: Transaction
}): Promise<{ qtyAvailable: number; status: string }> {
  const { batchId, qty, transaction } = opts
  const batch = await InvBatch.findByPk(batchId, { transaction, lock: transaction ? true : undefined })
  if (!batch) throw new NotFoundError('Batch not found')

  const available = Number(batch.qtyAvailable || 0) + qty
  const received = Number(batch.qtyReceived || 0)
  let status = batch.status ?? 'AVAILABLE'
  let category = batch.category
  if (['CONSUMED', 'PARTIALLY_CONSUMED'].includes(status)) {
    status = available < received ? 'PARTIALLY_CONSUMED' : 'AVAILABLE'
    category = 'available'
  }
  await batch.update({ qtyAvailable: available, status, category, updatedAt: new Date() }, { transaction })
  if (batch.includePack) await restoreToPacks(batchId, qty, transaction)

  await InvBatchEvent.create({
    batchId,
    eventType: 'ADJUSTMENT',
    qty,
    performedBy: opts.performedBy,
    performedAt: new Date(),
    remarks: opts.remarks ?? null,
  }, { transaction })

  return { qtyAvailable: available, status }
}
